src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs

```diff
@@ -34,7 +36,54 @@ public ValueTask<int> ReadAsync(Memory<byte> memory)
         return new ValueTask<int>(read, 0);
     }
 
-    public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+    public async ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
```

Comment:
This needs to be optimized a bit more:
- If there's a single segment and it's less than 65K then there should be no async state machine overhead.
- Otherwise slice 65K chunks and write those (skipping empty segments if there's a bug).

Reply:
Just a clarification on the first point, I haven't found anywhere that the maximum chunk size should be 65KB, only that there should be a maximum of 64KB - 1 chunks allowed per calls. Is that correct?

Reply:
I think that's what you expressed last time. So I would have expected logic more along the lines of:

```C#
while (!data.IsEmpty)
{ 
   var chunk = data.Slice(0, Math.Min(maxSize, data.Length));
   await WriteImplAsync(chunk); // Skip empty memory sequences in this implementation to work around the bug
   data = data.Slice(maxSize);
}
```

Reply:
I think there could still be a misunderstanding.

If I understand the issue correctly, we don't have a `maxBytesPerWrite` we are aiming for. Instead, we have to respect the maximum number of chunks allowed by [WriteEntityChunks](https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/ihttpresponse-writeentitychunks-method), which is 65535. 
As far as I understand, there's no limit on the chunk size itself, so there's no `maxBytesPerWrite`, it depends on how we decide to chunk the data. The way it's done now in the method that calls the native `WriteEntityChunks` (`AsyncWriteOperationBase.WriteSequence`) is by creating one chunk per segment in the `ReadOnlySequence` it receives in input.  

What I'm doing, is just limiting the amount of segments in the `ReadOnlySequence` before passing it down, so that it ends up producing a call to the native function with a `ReadOnlySequence` of, at most, 65535 segments.

Reply:
I might have been confused by this then https://github.com/dotnet/aspnetcore/pull/41239#issuecomment-1102352437. It seems like it can be done at this layer by merging chunks together no?

Reply:
I didn't think about merging chunks, it might be a better solution, I'll try it, thank you.

Reply:
Also I don't think we want to merge it into a single chunk, but < 65536 seems doable 😄

Reply:
of course :-)

While I was reviewing the code, I remembered that I've already tried merging the chunks but it seems not to be possible.  

The HTTP_DATA_CHUNK expects a pointer to the memory containing the data to transfer and its size. 
Since we are iterating through the ReadOnlySequence to produce the chunks, we have segments that I don't think it's safe to assume are adjacent, so I'm not sure there is a safe and efficient way to merge the chunks.

Reply:
Why isn't this possible? It would require a copy of course but it should be possible...

Reply:
I agree, but I've never said it is not possible, I said I'm not sure there is a safe and efficient way to do it, implying that to make it safe we would need to copy.

So I think we are in front of two different solutions with different tradeoffs.  

- calling the function multiple time
- calling the function one time by merging the chunks (copy)

You know the domain far better than I do, so you might have an intuition based on solid experience of which tradeoff is better, in this case feel free to suggest which solution to adopt. 

If I had to decide, I would need to implement both and do some testing and then discuss the results with you because, again, I have a feeling we will need to accept a tradeoff either way and knowing which one is better it depends on a lot of factors, such as most common use cases, which I might not be aware of.   
I'm willing to go down this road but I think we need to discuss which parameter are important for testing (memory allocation is the first that comes to mi mind) and then discuss the results.

Let me know what you think is best.

Reply:
@gfr-g I recommend calling the function multiple times as @davidfowl demonstrated in https://github.com/dotnet/aspnetcore/pull/41239/files#r863952571.

Reply:
I read some more of the context, and you have to call multiple times if the data greater than 65,535 bytes, but @davidfowl is you can copy all the segments into a single buffer until that point. 

You don't even have to do that though, can use [ReadOnlySequence.Slice(0, 65535)](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1.slice?view=net-6.0#system-buffers-readonlysequence-1-slice(system-int32-system-int32) to get things down to size.

The key hear is not to regress performance in the common case when the write is less than 65,536 bytes. So if the length of the `ReadOnlySequence<byte>` is less than that, it should pass through to WriteImplAsync directly. We should not even call an `async` method until we know there are more than 65,535 bytes because that creates a state machine (even though it's a struct for `ValueTask`.

Reply:
@davidfowl told me offline  the limit is on segments not bytes. We could make `WriteImplAsync`'s `ValueTask` an out parameter and then use the inline return value to indicate if there is more to write. That way we fast path the vast majority of writes without `await`ing anything in our code on every write.

Reply:
@halter73 I'm glad we are all finally on the same page regarding the root of the problem we are trying to solve.

I will think more about your suggestion, it's been a while since I looked at the code.

Reply:
I think the suggestion @halter73 is proposing requires some work on the `AsyncIOEngine`, while there might be a simpler solution to get rid of the awaits, something similar to the code structure proposed by [@davidfowl in this comment](https://github.com/dotnet/aspnetcore/pull/41239#discussion_r863952571), but adapted so that it slices based on segments borders.  

It's basically the original solution, calling the native function multiple times, while avoiding awaits as suggested.   

I'm working on it, hopefully I will be able to post an update soon. There are a few things I want to try to improve even though it doesn't seem likely there's an easy solution, like an efficient way to discard the empty segment produced as a consequence of the bug that appears when slicing a `ReadOnlySequence`.